drv/stm32h7-sprot-server/src/main.rs

```diff
@@ -0,0 +1,593 @@
+                                            // with a retry.
+                                            errcode = code;
+                                            break;
+                                        },
+                                    }
+                                }
+                            }
+                        },
+                        MsgType::ErrorRsp => {
+                            errcode = MsgError::Unknown;
+                            continue;
+                        },
+                        _ => {
+                            // All of the non-error message types are ok here.
+                            // XXX which length gets returned???
+                            return result;
+                        },
+                    }
+                },
+            }
+        }
+        Err(errcode)
+    }
+
+    /// Retrieve low-level RoT status
+    fn do_status(&mut self) -> Result<Status, MsgError> {
+        self.txmsg.init(MsgType::StatusReq, 0);
+        self.txmsg.set_crc(); // XXX enqueue(EnqueueBuf::Empty) would set the CRC.
+        self.do_send_recv(TIMEOUT_QUICK)?;
+        if_chain! {
+            if let Some(buf) = self.rxmsg.payload_buf();
+            // TODO: Do we need to allow for unligned?
+            if let Some(status) = LayoutVerified::<_, Status>::new(buf);
+            then {
+                return Ok(*status);
+            }
+        }
+        Err(MsgError::BadMessageLength)
+    }
+
+    /// Clear the ROT_IRQ and the RoT's Tx buffer by toggling the CSn signal.
+    /// ROT_IRQ before and after state is returned for testing.
+    fn do_pulse_cs(
+        &mut self,
+        delay: u64,
+        delay_after: u64,
+    ) -> Result<SpRotPulseStatus, MsgError> {
+        let rot_irq_begin = self.is_rot_irq_asserted();
+        self.spi
+            .lock(CsState::Asserted)
+            .map_err(|_| MsgError::CannotAssertCSn)?;
+        if delay != 0 {
+            hl::sleep_for(delay);
+        }
+        self.spi.release().unwrap_lite();
+        if delay_after != 0 {
+            hl::sleep_for(delay); // TODO: make this a 2nd parameter?
+        }
+        let rot_irq_end = self.is_rot_irq_asserted();
+        let status = SpRotPulseStatus {
+            rot_irq_begin: u8::from(rot_irq_begin),
+            rot_irq_end: u8::from(rot_irq_end),
+        };
+        Ok(status)
+    }
+
+    fn is_rot_irq_asserted(&mut self) -> bool {
+        self.sys.gpio_read(ROT_IRQ).unwrap_lite() == 0
+    }
+
+    // Poll ROT_IRQ until asserted (true) or deasserted (false).
+    fn wait_rot_irq(&mut self, desired: bool, max_sleep: u32) -> bool {
+        let mut slept = 0;
+        while self.is_rot_irq_asserted() != desired {
+            if slept == max_sleep {
+                ringbuf_entry!(Trace::RotReadyTimeout);
+                return false;
+            }
+            hl::sleep_for(1);
+            slept += 1;
```

Comment:
IIRC `hl::sleep_for()` only promises that it sleeps _at least as long as_ the argument we pass it, so this could potentially end up sleeping for much longer than `max_sleep`. It would be a bit more complex, but should we convert `max_sleep` to a deadline on entry to this function and break based on how much time we've actually slept?

Reply:
Sleeping longer affects performance, but is tolerable. To tune this properly, we need to know the LPC55 interrupt response times. They all should be sub-millisecond, but the variability remains unknown.
There is certainly room for improvement. Retries by the SP have worked well so far.